processing/src/main/java/io/druid/query/groupby/epinephelinae/AbstractBufferGrouper.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.groupby.epinephelinae;
+
+import com.google.common.base.Supplier;
+import com.google.common.primitives.Ints;
+import io.druid.java.util.common.IAE;
+import io.druid.java.util.common.logger.Logger;
+import io.druid.query.aggregation.AggregatorFactory;
+import io.druid.query.aggregation.BufferAggregator;
+
+import java.nio.ByteBuffer;
+
+public abstract class AbstractBufferGrouper<KeyType> implements Grouper<KeyType>
+{
+  protected static final int HASH_SIZE = Ints.BYTES;
+  protected static final Logger log = new Logger(AbstractBufferGrouper.class);
+
+  protected final Supplier<ByteBuffer> bufferSupplier;
+  protected final KeySerde<KeyType> keySerde;
+  protected final int keySize;
+  protected final BufferAggregator[] aggregators;
+  protected final int[] aggregatorOffsets;
+  protected final int bufferGrouperMaxSize; // Integer.MAX_VALUE in production, only used for unit tests
+
+  protected float maxLoadFactor;
+  protected int initialBuckets;
+  protected int bucketSize;
+
+  protected ByteBufferHashTable hashTable;
+  protected ByteBuffer hashTableBuffer; // buffer for the entire hash table (total space, not individual growth)
+
```

Comment:
can you please add comment on why above variables are not final? (I think because you can reset() and reuse the same Grouper )